### ISSUE TITLE
Spurious warning when running with Ruby 3.2: "undefining the allocator of T_DATA class LibXML::XML::Writer"

### ISSUE THREAD

**Issue by jcsahnwaldt** [xml4r/libxml-ruby#198]
When I use `LibXML::XML::Writer` in Ruby 3.2, I get `warning: undefining the allocator of T_DATA class LibXML::XML::Writer` on stderr. The reason seems to be [this change](https://bugs.ruby-lang.org/issues/18007) in Ruby. See e.g. https://github.com/swig/swig/pull/2258 and https://github.com/vmg/redcarpet/pull/721 on how to fix this.

**Comment by jcsahnwaldt** [comment]
I think adding the following line in https://github.com/xml4r/libxml-ruby/edit/master/ext/libxml/ruby_xml_writer.c should fix this:

```
rb_undef_alloc_func(cXMLWriter);
```

I could create a pull request, but I don't really know what the call does, and I'm not sure where to put it. Right after `cXMLWriter = rb_define_class_under(...)`? Or maybe after `rb_undef_method(CLASS_OF(cXMLWriter), "new")`?

**Comment by jcsahnwaldt** [comment]
I'd expect that some of the [other uses of `rb_define_class_under`](/xml4r/libxml-ruby/search?q=rb_define_class_under) will produce the same warning under Ruby 3.2 and will need a similar fix, but I haven't tested them.

**Comment by cfis** [comment]
That's interesting. So my understanding was calling `rb_undef_alloc_func` meant you can't create instances of that class - so any easy way to prevent someone creating an object that they should not because its created in C.  For classes where you do want to allow object creation from Ruby then you almost always have to define an alloc function. Guess my understanding wasn't quite right.

Anyway, so yes, an MR like you propose sounds good.

**Comment by cfis** [comment]
`rb_undef_alloc_func` removes the default alloc function which means you cannot call the new method anymore. So you can't create new instances of the class. Instead some other code in the extension needs to create them instead.

**Comment by jcsahnwaldt** [comment]
**TODO**:

- Do more [libxml-ruby classes defined by `rb_define_class`](https://github.com/search?q=repo%3Axml4r%2Flibxml-ruby+rb_define_class&type=code) produce the `undefining the allocator` warning?
- Is there a better way to find these classes?

**Comment by cfis** [comment]
Any class that has a new method that can be called from Ruby needs to have an allocator.

**Comment by jcsahnwaldt** [comment]
> Any class that has a new method that can be called from Ruby needs to have an allocator.

Sounds reasonable, but is it true for all cases? When I ran [my test script](https://gist.github.com/jcsahnwaldt/f4b859e2085d813ac0e5a718d61e257f), I saw that `LibXML::XML::Schema.new()` produces the `undefining the allocator` warning when it is first called. When the Ruby runtime prints the warning, it also calls `rb_undef_alloc_func()`. If I understand correctly, the `Schema` class then doesn't have an allocator anymore. But calling `LibXML::XML::Schema.new()` still works.

I guess the actual rule is a bit more complex, e.g. _"a class needs to have an allocator if it meets conditions X, Y and Z"_, where _X_ might be _"has a new method that can be called from Ruby"_, and _Y_ and _Z_ are qualifiers for special cases.

As I said, I don't really know what I'm doing here. But if I can contribute something useful without understanding all the details, I guess that's OK. :-)
